File: src/scripts/Lexing/context.ts
import type { Token } from "./lexer";
import { TokenType } from "./lexer";
import type { CalloutType } from "./Handlers/PseudoHTMLHandler";

export default class LexerContext {
    constructor(
        public input: string,
        public position: number = 0,
        public line: number = 1,
        public col: number = 1
    ) {}

    isEOF(): boolean {
        return this.position >= this.input.length;
    }

    peek(lookahead: number = 0): string {
        if (this.position + lookahead >= this.input.length) {
            return '';
        }
        return this.input[this.position + lookahead];
    }

    advance(count: number = 1): string {
        const char = this.peek();
        this.position += count;

        if (char === "\n") {
            this.line++;
            this.col = 1;
        } else {
            this.col += count;
        }
        return char;
    }

    matchString(str: string): boolean {
        for (let i = 0; i < str.length; i++) {
            if (this.peek(i) !== str[i]) {
                return false;
            }
        }
        return true;
    }

    createToken( type: TokenType, value: string, calloutType?: CalloutType, calloutTitle?: string, imgExternal?: string ): Token {
    return {
        type,
        value,
        position: { line: this.line, col: this.col - value.length },
        ...(calloutType && calloutTitle && { calloutType, calloutTitle }),
        ...(imgExternal && { imgExternal })
    }
}

}

File: rust-wasm/src/scripts/Lexing/context.rs
mod TokenType;

pub struct LexerContext<'a> {
    input: &'a str,
    position: u128,
    line: u128,
    col: u128
}

impl LexerContext {
    pub fn new(
        input: &str,
        position: u128 ,
        line: u128,
        col: u128
    ) -> Self { 
        Self { input, position, line, col }; 
    }

    fn is_eof(&self) -> bool {
        return self.position >= self.input.chars().count();
    }

    fn peek(&self, lookahead: u128) -> String {
        if (self.position + lookahead >= self.input.chars().count()) {
            return "";
        }
        return self.input
            .chars()
            .nth((self.position + lookahead) as usize)
    }

    fn advance(&self, count: Option<u64>) {
        let char = self.peek();
        let countint = count.unwrap_or(1);

        self.position = self.position + countint;

        if (char == "\n") {
            self.line = self.line + 1;
            self.col = 1;
        } else {
            self.col = self.col + countint;
        }

        return char;
    }

    fn create_token(&self, ttype: TokenType, value: String, )
}